pkg/timestream/macros.go

```diff
@@ -1,98 +1,106 @@
+	// sort macro keys longest first, so shorter keys don't clobber longer keys
+	// they're a prefix of
+	// TODO: consider refactoring sqlutil.Interpolate to be more generic and using that
+	macroKeys = maps.Keys(macroFuncs)
+	slices.SortFunc(macroKeys, func(a, b string) int { return cmp.Compare(-len(a), -len(b)) })
+}
 
-	// Simple Macros
-	txt = replaceOrDefault(txt, "$__database", query.Database, settings.DefaultDatabase)
-	txt = replaceOrDefault(txt, "$__table", query.Table, settings.DefaultTable)
-	txt = replaceOrDefault(txt, "$__measure", query.Measure, settings.DefaultMeasure)
+func macroTimeFilter(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
+	from := model.TimeRange.From.UnixNano() / 1e6
+	to := model.TimeRange.To.UnixNano() / 1e6
 
-	timeFilterExp, err := regexp.Compile(timeFilter)
-	if err != nil {
-		return txt, err
-	}
+	replacement := fmt.Sprintf("time BETWEEN from_milliseconds(%d) AND from_milliseconds(%d)", from, to)
+	return replacement, nil
+}
 
-	if timeFilterExp.MatchString(txt) {
-		timeRange := query.TimeRange
-		from := int64(timeRange.From.UnixNano() / 1e6)
-		to := int64(timeRange.To.UnixNano() / 1e6)
+func macroTimeFrom(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
+	return fmt.Sprintf("%d", model.TimeRange.From.UnixNano()/1e6), nil
+}
 
-		replacement := fmt.Sprintf("time BETWEEN from_milliseconds(%d) AND from_milliseconds(%d)", from, to)
-		txt = timeFilterExp.ReplaceAllString(txt, replacement)
-	}
+func macroTimeTo(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
+	return fmt.Sprintf("%d", model.TimeRange.To.UnixNano()/1e6), nil
+}
 
-	if strings.Contains(txt, timeFromStr) {
-		timeRange := query.TimeRange
-		from := int64(timeRange.From.UnixNano() / 1e6)
-		replacement := fmt.Sprintf("%d", from)
-		txt = strings.ReplaceAll(txt, timeFromStr, replacement)
+func macroInterval(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
+	if model.Interval.Milliseconds() == 0 {
+		return "", fmt.Errorf("invalid interval: %dns", model.Interval.Nanoseconds())
 	}
+	return fmt.Sprintf("%dms", model.Interval.Milliseconds()), nil
+}
 
-	if strings.Contains(txt, timeToStr) {
-		timeRange := query.TimeRange
-		to := int64(timeRange.To.UnixNano() / 1e6)
-		replacement := fmt.Sprintf("%d", to)
-		txt = strings.ReplaceAll(txt, timeToStr, replacement)
+func macroIntervalRaw(model models.QueryModel, _ models.DatasourceSettings) (string, error) {
+	if model.Interval.Milliseconds() == 0 {
+		return "", fmt.Errorf("invalid interval: %dns", model.Interval.Nanoseconds())
 	}
+	return fmt.Sprintf("%d", model.Interval.Milliseconds()), nil
+}
 
-	if strings.Contains(txt, intervalRawStr) {
-		replacement := fmt.Sprintf("%d", query.Interval.Milliseconds())
-		txt = strings.ReplaceAll(txt, intervalRawStr, replacement)
-	}
+func macroNow(_ models.QueryModel, _ models.DatasourceSettings) (string, error) {
+	now := time.Now().UnixNano() / 1000 // ms
```

Comment:
This reminded me about the existence of `UnixMill()`, which is even clearer!

Reply:
Ah, even better - it turns out the `now_ms` has been broken since it was introduced, and the test for it too. It's been returning time in microseconds rather than milliseconds. The [comparison it's using](https://github.com/grafana/timestream-datasource/blob/ef38963482627cb676687aea4041ecf6f8f7328a/pkg/timestream/macros_test.go#L95-L98) always returns `true`.